runtime/lua/vscode-neovim/api.lua

```diff
@@ -1,133 +1,170 @@
-local api = vim.api
-local fn = vim.fn
--- used to execute vscode command
-local command_event_name = "vscode-command"
--- used for extension communications
-local plugin_event_name = "vscode-neovim"
-
 local M = {}
 
--- send commands to vscode
-function M.notify(command, ...)
-  return vim.rpcnotify(vim.g.vscode_channel, command_event_name, command, { ... })
-end
-
-function M.call(command, ...)
-  return vim.rpcrequest(vim.g.vscode_channel, command_event_name, command, { ... })
-end
-
--- send commands to vscode extension
-function M.notify_extension(command, ...)
-  return vim.rpcnotify(vim.g.vscode_channel, plugin_event_name, command, { ... })
-end
-
-function M.call_extension(command, ...)
-  return vim.rpcrequest(vim.g.vscode_channel, plugin_event_name, command, { ... })
-end
-
--- send command to vscode with range (line or char). [1, 1]-based.
-function M.call_range(command, line1, line2, leaveSelection, ...)
-  return M.call_extension("range-command", command, "V", line1, line2, 1, 1, leaveSelection, { ... })
-end
+------- Requests -------
 
-function M.notify_range(command, line1, line2, leaveSelection, ...)
-  return M.notify_extension("range-command", command, "V", line1, line2, 1, 1, leaveSelection, { ... })
-end
+local REQUEST_SATE = {
+  id = 0,
+  callbacks = {},
+}
 
-function M.call_range_pos(command, line1, line2, pos1, pos2, leaveSelection, ...)
-  return M.call_extension("range-command", command, "v", line1, line2, pos1, pos2, leaveSelection, { ... })
+local function add_callback(callback)
+  REQUEST_SATE.id = REQUEST_SATE.id + 1
+  REQUEST_SATE.callbacks[REQUEST_SATE.id] = callback
+  return REQUEST_SATE.id
 end
 
-function M.notify_range_pos(command, line1, line2, pos1, pos2, leaveSelection, ...)
-  return M.notify_extension("range-command", command, "v", line1, line2, pos1, pos2, leaveSelection, { ... })
+---Invoke the callback, called by vscode
+---@param id number callback id
+---@param result any resutl
+---@param is_error boolean is error
+function M.invoke_callback(id, result, is_error)
+  vim.schedule(function()
+    local callback = REQUEST_SATE.callbacks[id]
+    REQUEST_SATE.callbacks[id] = nil
+    if callback then
+      if is_error then
+        callback(result, nil)
```

Comment:
I think it should be (res, err), to match js (resolve, reject).

Reply:
I hope users will pay attention to error handling, just like pcall.

Reply:
Is that worth the strange ordering though? Most users will be writing simple keybindings, and don't really care about error handling.

Reply:
If you use simple keibindings, you won't use the callback parameter.